### PR TITLE
Make salt configurable. Write catalog file correctly if documents are not rewritten.

### PR DESCRIPTION
**Pull request by danielballan**
The default behavior, unchanged by this PR, is to generate a random salt per export operation so that the target directories for the data are unique. Now you have the option to specify a `--salt` manually, which means you can export into deterministic directories, handling any conflicts that arise during file transfer as you see fit.

Also, when `--no-documents` is set, we know that we cannot rewrite the `root` in the Resource document because we aren't writing out any documents, so we leave the key in  `root_map` alone.

### CODE REVIEW

File: databroker_pack/_pack.py [hunk 1]
```diff
@@ -306,15 +318,29 @@ def export_run(
...
+                            # the data if we plan to copy it.
+                            # - unique_id is unique to this (root, salt)
+                            # combination and used to place the data in a
+                            # unique location.
+                            key = (root_in_document, root, unique_id)
+                            files[key].update(run.get_file_list(doc))
+                        if not dry_run:
+                            # Replace root with a unique ID before serialization.
+                            # We are overriding the local variable name doc here
+                            # (yuck!) so that serializer(name, doc) below works on
```
Comment (tacaswell) [inline databroker_pack/_pack.py:340]: You do the same on L311 without editorial comment.

**Comment by tacaswell** [comment]
The mixing of `dry_run` and `nodocuments` is a bit confusing (as I think it is mixing something that was a testing / patch up functionality with something we need for running a single mongo that is accessed from 2 file systems (with mirrored payloads)).

**Comment by tacaswell** [comment]
over all seems :+1:

File: databroker_pack/_pack.py (hunk 1 above)
Comment (danielballan) [reply databroker_pack/_pack.py:340]: True. As we've said elsewhere, it sure looks like `DocumentRouter` needs a callback registry or a streamz dependency.

**Pull request merged**